### ISSUE TITLE
Warn that `v.into_iter().take().collect::<Vec<_>>()` might not release `v`'s memory

### ISSUE THREAD

**Issue by mcalancea** [rust-lang/rust-clippy#15753]
### What it does

It warns about the issue described [here](https://blog.polybdenum.com/2024/01/17/identifying-the-collect-vec-memory-leak-footgun.html).

TLDR: `v.into_iter()...collect::<Vec<_>>()` can in general be optimized to run in-place and use `v`'s memory. But this means that, for example, `v.into_iter().take(k).collect::<Vec<_>>()` will have `len() = k` and `capacity() = v.len()`. See [playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=194fa11c1f9d27900574684618ab0e7e).

Clippy should warn about such patterns (the degree of generality remains open) and suggest applying `shrink_to_fit` to the result or using different methods of extracting the desired values.

### Advantage

The problematic code may result in unexpectedly high memory usage. The suggested changes (or even just raising awareness about this behavior) would prevent such scenarios.

### Drawbacks

It's difficult to know exactly when the in-place optimization applies, so the user should manually check if the suggested change is needed in their case.

I guess it might be annoying for those who know the internal iterators well and expect this behavior.

### Example

```rust
    let b: Vec<_> = a.into_iter().take(5).collect();
```

Could be written as:

```rust
    let b: Vec<_> = {
        let mut b = a.into_iter().take(5).collect();
        b.shrink_to_fit();
   };
```
Although there are probably better options.

**Comment by lukaslueg** [comment]
[Relevant](https://github.com/rust-lang/rust/issues/120091), which resulted in https://github.com/rust-lang/rust/pull/120355
